File: source/vehicle_detector.cpp
#include <iostream>

#include "vehicle_detector.h"

using namespace std;


VehicleDetector::VehicleDetector()
{
    net = cv::dnn::readNet("models/yolov3.weights", "models/yolov3.cfg");
    net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
    net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);

    scalefactor = 1 / 255.0;
    mean = cv::Scalar(127.5, 127.5, 127.5);
    conf_threshold = 0.9;
    nms_threshold = 0.4;

    model = cv::dnn::DetectionModel(net);
    model.setInputParams(scalefactor, cv::Size(416, 416), mean, true);
}

VehicleDetector::~VehicleDetector()
{
    // delete ui;
}

void VehicleDetector::detect(cv::Mat& image, vector<Vehicle>& vehicles)
{
    bboxes.clear();
    centers.clear();
    class_ids.clear();
    confidences.clear();

    start_time = clock();

    model.detect(image, class_ids, confidences, bboxes, conf_threshold, nms_threshold);
    
    end_time = clock();
    cout << "vehicle detect time: " << float(end_time - start_time) / CLOCKS_PER_SEC << " s" << endl;

    for (int i = 0; i < bboxes.size(); i++)
    {
        vehicles.push_back(Vehicle(image, bboxes[i], class_ids[i]));
    }
}

File: python/source/vehicle_detector.py
import sys
import os
from functools import partial
from datetime import datetime
import time

import cv2
import numpy as np

from .functions import *
from .vehicle import Vehicle


class VehicleDetector:
    def __init__(self):
        self.net = cv2.dnn.readNetFromDarknet("models/yolov3.cfg", "models/yolov3.weights")
        self.net.setPreferableBackend(cv2.dnn.DNN_BACKEND_CUDA)
        self.net.setPreferableTarget(cv2.dnn.DNN_TARGET_CUDA)

        self.conf_threshold = 0.7
        self.nms_threshold = 0.45

        self.size = (416, 416)
        self.scalefactor = 1 / 255.0
        self.model = cv2.dnn_DetectionModel(self.net)
        self.model.setInputParams(self.scalefactor, self.size)

    def detect(self, image):
        vehicles = []
        class_ids, confidences, boxes = self.model.detect(image, self.conf_threshold, self.nms_threshold)
        for i in range(len(boxes)):
            vehicles.append(Vehicle(rect=boxes[i], class_id=class_ids[i], confidence=confidences[i]))

        return vehicles

